### PR TITLE
Initial creation of API

### PR DESCRIPTION
**Pull request by seanlanepgh**

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            throw new Exception\Unauthorized($message, 400, $e);
+        }
+        $decode = json_decode((string) $authenticate->getBody(), true);
+        $this->setToken($decode['token']);
+    }
+
+    /**
+     * set token
+     * @param string $token Token must be set to have access
+     *  to the api and is only valid for a hour
```
Comment (williamrenfrew) [inline src/Client.php]: 'an hour'

File: src/Client.php [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    public function checkToken() : void
+    {
+        if ($this->token === null || $this->token->isExpired()) {
+            $this->authenticate();
+        }
+    }
+
+    /**
+     * This request function handles all requests for the api
+     * @param  string $type     Stores the type of HTTP Request e.g GET ,POST
```
Comment (williamrenfrew) [inline src/Client.php]: 'Sets'

File: src/Client.php [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            throw $exception;
+        }
+
+        $res = (array)json_decode((string) $res->getBody(), true);
+
+        return $res;
+    }
+
+    /**
+     *  A function that handles the creation of GET  Request
```
Comment (williamrenfrew) [inline src/Client.php]: A function that handles the creation of a GET Request

File: src/Client.php [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+     * @param  string $endpoint  A endpoint used to create a request
```
Comment (williamrenfrew) [inline src/Client.php]: An

File: src/ListResult.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    private $targetClass;
+    private $endpoint;
+    private $params;
+    private $pageSize = 10;
+    private $currentPagePos = null;
+    private $currentRecordPos = null;
+    private $currentRecordSet = [];
+
+    /**
+     * This constructor is used to build the ListResult
```
Comment (williamrenfrew) [inline src/ListResult.php:20]: Descriptors reqd

**Review by williamrenfrew — CHANGES_REQUESTED** [review]

File: src/ListResult.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+     * @param Client $client      [description]
```
Comment (willpower232) [inline src/ListResult.php]: complete descriptions or remove `[description]`

File: src/ListResult.php [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    public function rewind() : void
+    {
+        $this->page(1);
+    }
+
+    /**
+     * Get the current page
+     * @return void
+     */
+    public function current()
```
Comment (willpower232) [inline src/ListResult.php:109]: probably not return void?

File: src/Client.php [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+{
+    private $http_client;
+    private $token;
+    private $config;
+
+    /**
+     * construct function that builds the client class
+     * @param array $config creditsafe configuration
+     */
+    public function __construct($config = [])
```
Comment (willpower232) [inline src/Client.php]: missing type declaration

File: src/ListResult.php [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        }
+        return array_merge($this->params, $pageParams);
+    }
+
+    /**
+     * Get Page
+     * @param  int    $num  Passes the page number
+     * @return array  Returns  the current Record Set
+     */
+    public function page(int $num)
```
Comment (willpower232) [inline src/ListResult.php]: missing return type declaration

File: src/ListResult.php [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    public function current()
+    {
+        return $this->currentRecordSet[$this->key()];
+    }
+
+    /**
+     * Get the next page
+     * @return void
+     */
+    public function next()
```
Comment (willpower232) [inline src/ListResult.php]: missing return type declaration

File: src/ListResult.php [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        if ($this->currentRecordPos > $this->maxPageRecord) {
+            $this->page($this->currentPagePos+1);
+        }
+    }
+
+    /**
+     * Get the key of a position
+     * @return void
+     */
+    public function key()
```
Comment (willpower232) [inline src/ListResult.php]: missing return type declaration

File: src/ListResult.php [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        return $this->currentRecordPos - ($this->pageSize * ($this->currentPagePos-1));
+    }
+
+    /**
+     * Check if a position is valid
+     * @return bool
+     */
+    public function valid()
```
Comment (willpower232) [inline src/ListResult.php]: missing return type declaration

File: src/Client.php [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        $this->checkToken();
+
+        $guzzleArgs =  [
+             'headers' => [
+                 'Authorization' => (string) $this->token
+             ],
+         ];
+
+        if ($type == 'GET') {
+            $guzzleArgs ['query']= $params ;
```
Comment (williamrenfrew) [inline src/Client.php]:
```suggestion
            $guzzleArgs ['query'] = $params;
```

File: src/Client.php [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+        } else {
+            $guzzleArgs['json']= $params ;
```
Comment (williamrenfrew) [inline src/Client.php]:
```suggestion
            $guzzleArgs['json'] = $params;
```

**Review by williamrenfrew — CHANGES_REQUESTED** [review]
Some general formatting required too

File: src/Models/CompanySearchResult.php [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     *
+     * @return DateTime Get Date Of Latest Change
+     */
+    public function getDateOfLatestChange() : \DateTime
+    {
+        return  new \DateTime($this->dateOfLatestChange);
+    }
+
+    public function get()
```
Comment (willpower232) [inline src/Models/CompanySearchResult.php]: missing return type declaration

File: tests/ClientTest.php [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+namespace SynergiTech\Creditsafe\Tests;
+
+use SynergiTech\Creditsafe\Client;
+
+class ClientTest extends \SynergiTech\Creditsafe\Tests\Base
+{
+    /**
+     * @dataProvider providerAuthorizationToken
+     */
+    public function testAuthorizationSuccess($guzzle)
```
Comment (willpower232) [inline tests/ClientTest.php:12]: missing function and return type declarations

**Review by willpower232 — CHANGES_REQUESTED** [review]
great start, please have another look through all the files and make sure you're being consistent, particularly with white space usage. Hopefully atom can sort out most of it but do have another look after too.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+### Access countries and their codes
+```php
+$creditsafe->countries()->access($params = []);
+```
+
+### Search criteria using country code
+```php
+
+$creditsafe->companies()->searchCriteria(['countries'=>'GB']);
```
Comment (Josh-G) [inline README.md]:
```suggestion
$creditsafe->companies()->searchCriteria(['countries' => 'GB']);
```

File: README.md [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+
+```
+### Company search pagination
+```php
+$search = $creditsafe->companies()->search(['countries'=>'GB', 'name' => 'GOOGLE UK LIMITED']);
```
Comment (Josh-G) [inline README.md]:
```suggestion
$search = $creditsafe->companies()->search(['countries' => 'GB', 'name' => 'GOOGLE UK LIMITED']);
```

File: src/ListResult.php [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        $this->targetClass = $targetClass;
+        $this->endpoint = $endpoint;
+        $this->params = $params;
+        $this->page(1);
+    }
+
+    /**
+     * Set PageSize
+     * @param  int  $size Stores the PageSize
+     * @return self   Returns PageSize
```
Comment (Josh-G) [inline src/ListResult.php]:
```suggestion
     * @return self
```
wrong description, probably doesn't need one

File: src/ListResult.php [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+        $this->pageSize = $size;
+        $this->rewind();
+        return $this;
+    }
+
+    /**
+     * Get  Params
+     * @return array Return Params
+     */
+    private function getParams(): array
```
Comment (Josh-G) [inline src/ListResult.php]:
```suggestion
    private function getParams() : array
```

File: src/Models/Company.php [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+    private $commentaries;
+    private $history;
+    private $currentDirectors;
+    private $previousDirectors;
+    private $creditScore;
+    private $mortgageSummary;
+    private $mortgages;
+    private $negativeInfo;
+    private $financialStatements;
+    /**
```
Comment (Josh-G) [inline src/Models/Company.php:33]:
```suggestion

    /**
```

File: src/Models/Company.php [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+            $history['date'] = new \DateTime($history['date']);
+            return $history;
+        }, $companyDetails['report']['additionalInformation']['companyHistory'] ?? []);
+
+        $this->commentaries = $companyDetails['report']['additionalInformation']['commentaries'];
+        $this->creditScore = new Company\CreditScore($this, $companyDetails['report']['creditScore']);
+        $this->mortgageSummary = $companyDetails['report']['additionalInformation']['mortgageSummary'];
+        $this->mortgages = $companyDetails['report']['additionalInformation']['mortgageDetails'] ?? null;
+        $this->negativeInfo = $companyDetails['report']['negativeInformation'];
+        $this->rawDetails = $companyDetails;
```
Comment (Josh-G) [inline src/Models/Company.php:76]: add this to private properties above

File: src/Models/Company/Director.php [hunk 23]
```diff
@@ -0,0 +1,106 @@
+<?php
+
+
```
Comment (Josh-G) [inline src/Models/Company/Director.php]:
```suggestion
```

File: src/Models/Company/Director.php [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+     * @param bool $current            Boolean used to indicate if the director is a current or previous director
+     */
+    public function __construct(Company $company, array $directorDetails, bool $current)
+    {
+        $this->company = $company;
+        $this->directorDetails = $directorDetails;
+        //Add if statement to check if position exists
+        if (isset($this->directorDetails['positions']['0']['dateAppointed'])) {
+        //Can't assign datetime due to previousDirectors not having dateAppointed
+            $this->directorDetails['positions']['0']['dateAppointed'] = new \DateTime($directorDetails['positions']['0']['dateAppointed']);
```
Comment (Josh-G) [inline src/Models/Company/Director.php]: we should be applying this to every director position, not just the first

File: src/Models/Company/FinancialStatement.php [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Function constructs the FinancialStatement Class
+     * @param Company $company          Used to store a company data in the Director Class
+     * @param array  $statementDetails  Financial Statement Data that needs to be stored in the FinancialStatement Class
+     */
+    public function __construct(Company $company, $statementDetails)
+    {
+        $this->company = $company;
+        $this->statementDetails = $statementDetails;
+        var_dump($statementDetails);
```
Comment (Josh-G) [inline src/Models/Company/FinancialStatement.php]: this needs to come out

File: tests/ClientTest.php (hunk 16 above)
Comment (Josh-G) [reply tests/ClientTest.php:12]: i don't think we should apply strict types in test method signatures

**Review by Josh-G — CHANGES_REQUESTED** [review]
a couple of blockers

**Review by williamrenfrew — APPROVED** [review]

**Review by williamrenfrew — APPROVED** [review]

**Pull request merged**